### PR TITLE
nodejs plugin: add support for yarn.

### PR DESCRIPTION
**Pull request by sergiusens**
LP: #1639837

### CODE REVIEW

File: snapcraft/plugins/nodejs.py [hunk 1]
```diff
@@ -139,6 +163,41 @@ def _npm_install(self, rootdir):
...
+        else:
+            yarn_add = yarn_cmd + ['add']
+        for pkg in self.options.node_packages:
+            self.run(yarn_add + [pkg] + flags, cwd=rootdir)
+
+        # local packages need to be added as if they were remote, we
+        # remove the local package.json so `yarn add` doesn't pollute it.
+        if os.path.exists(self._source_package_json):
+            with contextlib.suppress(FileNotFoundError):
+                os.unlink(os.path.join(rootdir, 'package.json'))
```
Comment (kyrofa) [inline snapcraft/plugins/nodejs.py]: Ah ha, clever (and annoying).

**Comment by sergiusens** [comment]
yay, my latest fix does the trick 
`test_nodejs_plugin.NodeJSPluginTestCase.test_build_with_run_commands (yarn) ... ok`

But but, we seem to have reached a travis timeout.

File: snapcraft/plugins/nodejs.py [hunk 2]
```diff
@@ -101,20 +111,31 @@ def get_build_properties(cls):
...
     def __init__(self, name, options, project):
         super().__init__(name, options, project)
+        self._source_package_json = os.path.join(
+            os.path.abspath(self.options.source), 'package.json')
         self._npm_dir = os.path.join(self.partdir, 'npm')
         self._nodejs_tar = sources.Tar(get_nodejs_release(
             self.options.node_engine, self.project.deb_arch), self._npm_dir)
+        if self.options.node_package_manager == 'yarn':
+            logger.warning(
+                'EXPERIMENTAL: use of yarn to manage packages is experimental')
```
Comment (sergiusens) [inline snapcraft/plugins/nodejs.py]: I hope EXPERIMENTAL carries that connotation

File: snapcraft/plugins/nodejs.py (hunk 1 above)
Comment (sergiusens) [inline snapcraft/plugins/nodejs.py]: Yes and I had to do it differently as well for the `run` part

File: snapcraft/plugins/nodejs.py [hunk 3]
```diff
@@ -35,8 +35,12 @@
       (list)
       A list of targets to `npm run`.
       These targets will be run in order, after `npm install`
+    - node-package-manager
```
Comment (kyrofa) [inline snapcraft/plugins/nodejs.py]: This was resolved on rocket: these are namespaced by plugin so we can tell that they're not core options.

**Pull request merged**